[DOC] Governance update (#615)
ME-ICA/tedana: docs/governance.rst

Overview
--------

Tedana is a relatively small open source project that requires specialized
knowledge in multiple domains. This leads to several challenges. No one
person on the current tedana development team has a combination of 
available time plus expertise in collaborative software development, MRI
physics, and advanced data processing methods to assume a primary project
leader role. Even if such a person was interested, it may not benefit the
project to overly rely on the existence of one person. We developed the
following system with several goals in mind:

- Grow the community.
- Strive for consensus.
- Provide a path for when consensus cannot be achieved.
- Minimize the administrative burden.
- Maximize the `bus factor`_ of the project.
- Acknowledge the leadership and time multiple people contribute to our
  community without demanding more time than any individual can offer.
  Dividing leadership responsabilies into multiple smaller roles also
  makes it easier to encourage new people to take on a leadership role
  without fearing that too much work will be required of them.

> - Openness is a priority:

Reviewer: ```suggestion
- Openness as a priority:
```

  - Promote open discussions.
  - Openness is critical to building trust with the broader community
  - Openness provides a mechanism for non-leaders to identify and address

>     blindspots

Reviewer: oversights or mistakes.

  - Openness provides a smoother transition to onboard future leaders
  - Leadership meetings should be open and notes should be shared unless

>     there are discusssions about senisitive personal matters.

Reviewer: ```suggestion
    there are discussions about sensitive personal matters.
```

This governance structure is a work-in-progress. We welcome both people
who want to take on a leadership role as well as ideas for our to improve
this structure.

Leadership
----------

Contributor
  A contributor is someone who has made a contribution to tedana. A
  contribution, can be code, documentation, or conceptual. All contributors are
  listed in the `all-contributors file`_. The community decides on the content
  of this file using the same process as any other change to the `Repository`_
  (see below) allowing the meaning of "Contributor" to evolve independently of
  the Decision-making rules. Contributors also have the option to be added to
  the Zenodo file which may be used for authorship credit for tedana.
    

Maintainer

>   A Contributor responsible for the long term health of the project and the

Reviewer: ```suggestion
  A Contributor is responsible for the long term health of the project and the
```

  community. Maintainers have additional authority (see `Decision Making Process`_) helping them to
  resolve conflicts and increase the pace of the development when necessary.

>   Any maintainer can self-remove themselves. Any contributor can become a

Reviewer: ```suggestion
  Any maintainer can remove themselves. Any contributor can become a
```

>   maintainer by request and with the support of the majority of the current

Author: ```suggestion
  maintainer by request, or by nomination of a current maintainer,  and with the support of the majority of the current
```

>   maintainers. Current Maintainers:

Reviewer: ```suggestion
  maintainers, or by nomination of a current maintainer. Current Maintainers:
```

  +-----------------------------------+
  | `Dan Handwerker`_ (@handwerkerd)  |
  +-----------------------------------+
  | `Taylor Salo`_ (@tsalo)           |
  +-----------------------------------+
  | `Joshua Teves`_ (@jbteves)        |
  +-----------------------------------+
  | `Eneko Uruñuela`_ (@eurunuela)    |
  +-----------------------------------+
  | `Kirstie Whitaker`_ (@KirstieJane)|
  +-----------------------------------+

Steering committee
  The steering committee is made up of a subset of maintainers who help guide
  the project. Current Steering Committee members:

  +----------------------------------------------+
  | `Dan Handwerker`_ (@handwerkerd) (tentative) |
  +----------------------------------------------+ 
  | Your name here!                              |
  +----------------------------------------------+
  | Your name here!                              |
  +----------------------------------------------+
  | Your name here!                              |
  +----------------------------------------------+
  | Your name here!                              |
  +----------------------------------------------+

Focused Leadership Roles

>   We have identified key responsabilities or skills that help advance tedana

Reviewer: ```suggestion
  We have identified key responsibilities or skills that help advance tedana
```

>   development and created roles for each of these responsabilities. One

Reviewer: ```suggestion
  development and created roles for each of these responsibilities. One
```

  person can fill more than one role and more than one person can decide to

>   share or split the responsabilities of a role. Any contributor can proposed

Reviewer: ```suggestion
  share or split the responsibilities of a role. Any contributor can proposed
```

>   share or split the responsabilities of a role. Any contributor can proposed

Reviewer: ```suggestion
  share or split the responsibilities of a role. Any contributor can proposed
```

  the creation of new focused leadership roles.  A person can take on a
  leadership role without being a Maintainer or Steering Committee member

  - | Task manager & record keeper: `Dan Handwerker`_
    |   Helps write & keep track of notes from meetings
    |   Keeps track of issues or items that should be addressed
    |   Follows up with people who volunteered to address an item or alerts the broader community of known tasks that could use a volunteer
  - | MR physics leader `César Caballero-Gaudes`_
    |   Someone who can make sure calculations fit within our understanding of MR physics
    |   Someone who can either answer MRI physics questions related to multi-echo or direct people to where they can find answers

>   - | Processing algorithms leader `Eneko Uruñuela`_ (Decomposiiton) & `Dan Handwerker`_ (Metrics & Decision Tree)

Reviewer: ```suggestion
  - | Processing algorithms leader `Eneko Uruñuela`_ (Decompositon) & `Dan Handwerker`_ (Metrics & Decision Tree)
```

    |   If this is one person, that person should NOT be on the steering committee
    |   If this is more than one person, at least one should not be on the steering committee
    |   Ideal is someone who cares about tedana but DOESN’T know contributors well enough to say, ”Person X would never do that”

Changing leaders
````````````````
Any leader can remove themselves for a role at any time and open up a call for
a new self-nomination. Anyone can request to take on a leadership role at any
time. Once per year, there should be an explicit call to the larger contributor
community asking if anyone wants to self nominate for a leadership role. If
individuals cannot reach consensus on who steps back and who assumes new roles,
then a majority vote of contributors from the previous 3 years will assign
people to roles where there are conflicts.

If there are concerns with a tedana leader, any enforcer of the code of conduct
can ask anyone to step down from a leadership role. If a person refuses to step
down, then an enforcer of the code of conduct will consult with the other code
of conduct enforces. If they reach a concensus that a person shouldn't have a

> tedana leadership position, then they should be removed. Of a code of conduct

Reviewer 2: ```suggestion
tedana leadership position, then they should be removed. If a code of conduct
```

> enforcer as a conflict of interest, then the remaining code of conduct enforcers

Reviewer 2: ```suggestion
enforcer has a conflict of interest, then the remaining code of conduct enforcers
```

will identify someone without a conflict to include in deliberations.

Decision Making Process
-----------------------

The rules outlined below are inspired by the 
`decision-making rules for the BIDS standard <https://github.com/bids-standard/bids-specification/blob/master/DECISION-MAKING.md>`_, which in turn were inspired by the
`lazy consensus system used in the Apache Foundation <https://www.apache.org/foundation/voting.html>`_,
and heavily depend on the
`GitHub Pull Request review system <https://help.github.com/articles/about-pull-requests/>`_.

1. Potential modifications to the Repository should first be proposed via an
   Issue.

> 2. Every modification of the specification (including a correction of a typo,

Reviewer: ```suggestion
2. Every modification (including a correction of a typo,
```

>    adding a new Contributor, an extension adding support for a new data type,

Reviewer: ```suggestion
   adding a new Contributor, an extension
```

   or others) or proposal to release a new version needs to be done via a
   Pull Request (PR) to the Repository.
3. Anyone can open an Issue or a PR (this action is not limited to Contributors).
4. A PR is eligible to be merged if and only if these conditions are met:

   a) The PR features at least two `Reviews that Approve <https://help.github.com/articles/about-pull-request-reviews/#about-pull-request-reviews>`_
      the PR of which neither is the author of the PR. 
      The reviews should be made after the last commit in the PR (equivalent to 
      `Stale review dismissal <https://help.github.com/articles/enabling-required-reviews-for-pull-requests/>`_
      option on GitHub). If a second review requests minor changes after
      another reviewer approved the PR, the first review does not need
      to re-review.
   b) Does not feature any `Reviews that Request changes <https://help.github.com/articles/about-required-reviews-for-pull-requests/>`_.
      That is, if someone asked for changes, the PR should not be merged just
      because two other people approve it.

>    c) Is not a Draft PR. That is the PR author says it is ready for review.

Reviewer 2: ```suggestion
   c) Is not a Draft PR. That is, the PR author says it is ready for review.
```

   d) Passes all automated tests.

>    e) Is not proposing a new release

Reviewer: ```suggestion
   e) Is not proposing a new release.
```

   f) The steering committee has not added extra restrictions. For example, if
      a PR is a non-trival change, the steering committee can create a system
      to get feedback from more than just two reviewers before merging.
5. After consultation with contributors, the steering committee can decide
   to merge any PR - even if it's not eligible to merge according to Rule 4.
6. Anyone can Review a PR and request changes. If a community

>    member Requests changes they need to provide an explanation regarding what

Reviewer: ```suggestion
   member requests changes they need to provide an explanation regarding what
```

   changes should be made and justification of their importance. Reviews
   requesting changes can also be used to request more time to review a PR.

> 7. A reviewer who Requested changes can dismiss their own review, if

Reviewer: ```suggestion
7. A reviewer who requested changes can dismiss their own review, if
```

   they decide their requested changes are no longer necessary, or approve
   changes that address the issue underlying their change request.
8. If the author of a PR and a reviewer who requests changes cannot find a
   solution that would lead to: (1) The author closing the PR without merging
   (2) The reviewer accepting requested changes or (3) The dismissing their
   review, so that the PR can be approved and merged, then the disagreement
   will be resolved with a vote.
9. Rules governing voting:

   a) A vote can be triggered by any Maintainer, but only after 5 working days

>       from the time a Review Requesting Changes. If a PR has a disagreement

Reviewer: ```suggestion
      from the time a Review Requesting Changes is made. If a PR has a disagreement
```

>       that required a vote, there must be at least 15 days from the conclusion

Reviewer: Hm, on further reflection I'm not sure if I agree. This could easily let things drag out much more than they otherwise would. I'd be curious to hear what others think.

Author: I didn't change this from @tsalo's original draft. I suspect this makes sense for a larger project. I assume the spirit of this rule is to not have a PR jump from one contentious vote to another. I hope there are simpler solutions. For example, we allow the steering committee to change the rules so, if something is contentious, then can set up a system for resolving the contention rather than jumping from vote to vote.

Reviewer 2: There was a lot of discussion around this point in BIDS, and the thought was that a longer time period was necessary given the global nature of the project and the conflicting time commitments that so many members have.

That being said, we're at the point in BIDS where a vote basically never happens -- and I think that's (largely) a good thing. Votes are usually a sign that something in scope, etc, are unclear, if no one can agree and there's a need to poll the community. It also exhausts your contributors's time if they're being called for votes too often, whch means they have less bandwidth to actually contribute to the project.

I'm fine to copy the rules as-is with this understanding that votes are really the fall-back. If we're hitting them often, I think we need to take a deep look at where the confusion is arising and have a very clear discussion / decision on that.

>       of the first vote before another vote can be triggered.

Author: ```suggestion
      from the time a Review Requesting Changes. A PR can only have one 
      open vote at a time. If disagreements over a PR results in more than one
      vote, the Steering Committee has the authority to  create a voting process
      to help resolve disagreements in a more efficient and respectful manner.
```

Author: @emdupre & @jbteves I did propose this change based on several comments. I removed the 15 day delay and simply said we should never have more than one vote on a PR happening at one time and that, if a PR goes through more than one vote, the steering committee can decide how to create a resolution process. I think this get's at Elizabeth's point in her comment that, if there are repeated votes, the issue is due to the scope or explanation rather than the specific detail at hand. I think this is better than the arbitrary, but specific 15-days rule.
Does this look like a reasonable change to both of you?

Reviewer: I prefer this version, and think it's good.

>    a) A vote can be triggered by any Maintainer, but only after 5 working days
>       from the time a Review Requesting Changes is made. If a PR has a disagreement
>       that required a vote, there must be at least 15 days from the conclusion
>       of the first vote before another vote can be triggered.

Author: ```suggestion
   a) A vote can be triggered by any Maintainer, but only after 5 working days
     from the time a Review Requesting Changes is made. A PR can only have one 
      open vote at a time. If disagreements over a PR results in more than one
      vote, the Steering Committee has the authority to  create a voting process
      to help resolve disagreements in a more efficient and respectful manner.
```

Reviewer: I would agree with this. WDYT @emdupre @tsalo @dowdlelt ?

Reviewer 3: I think this works for me. Constrains things in a way that I think is productive.  Think I see an extra space between "authority to" and "create a voting"

Reviewer: No extra space if you go to edit it, maybe a weird rendering thing?

>    b) Only Maintainers can vote and each Maintainer gets one vote.

Reviewer 2: I'd prefer if this was Contributors rather than Maintainers.

Is there a reason to go with Maintainers here, @tsalo @handwerkerd ?

Author: I was keeping this as it was. The advantage of Maintainers is a that conflict is resolved by a clear number of people who are invested in the greater whole. The disadvantage is that there's no guarantee that Maintainers will understand a specific issue more than other contributors. As long as votes won't get slowed down waiting for a quorum, I'm find changing to Contributors.

Reviewer: I am also fine with changing to Contributors, particularly since it could be a vote on user-facing behavior.

Author: ```suggestion
   b) Only Contributors can vote and each Contributor gets one vote.
```

>    c) A vote ends after 7 working days or when all Maintainers have voted or

Author: ```suggestion
   c) A vote ends after 7 working days or when all Contributors have voted or
```

Author: One weirdness of giving all contributors a vote (instead of maintainers, is that all contributors will never abstain or vote so I assume a vote will always take 7 days. Not a big deal, but worth mentioning here.

Reviewer: I think we'll just have to be patient. Yes, we won't get all contributors to vote, but there's only ever rushes to make decisions during hackathons and hopefully we're not making votable decisions *at* a hackathon.

      abstained (whichever comes first).

>    d) A vote freezes the PR - no new commits or Reviews Requesting changes can

Reviewer: ```suggestion
   d) A vote freezes the PR - no new commits or Reviews Requesting Changes can
```

      be added to it while a vote is ongoing. If a commit is accidentally made
      during that period it should be reverted. Comments are allowed.

>    e) The quorum for a vote is five votes.

Reviewer 2: Just flagging that this might be high for our community, though I agree that 3 feels too low ! If we go with 5, we'll definitely need the longer (e.g. 15 day) time period.

Author: My sense is, if a vote it's called, it's likely after several weeks of discussion and attempts to resolve a conflict without a vote. 7 feels reasonable to me, but my opinion on this isn't particularly strong.

Reviewer 4: One other thing is that there's no good infrastructure for voting on GitHub, so it will always end up being a slow process. E.g., we don't have ways to automatically count votes, to set time limits on them, or to block pull requests based on the results of a vote. I think that's part of why BIDS has the voting system in their decision-making file, but it never gets used as described.

Author: If only Maintainers can merge pull requests and our maintainers are aware of these rules, we can block PRs during voting. Also, by definition, a vote will only happen if there is an unresolved comment from a reviewer, which would prohibit any random person from merging the PR.

I also like not describing the mechanics of voting in these rules because I think that flexibility will be useful. We know everyone who is allowed to vote, and, if we ever have a large enough number of voters such that a manual tally would be difficult, we'll have had to revise the bylaws well before that point. 

As for the practical mechanics, every vote could be an issue that lists the vote closing deadline, and people vote by commenting on that issue. I don't think we need to specify that in the docs, but it's just one option.

Reviewer: I believe that with a group this size, getting lost in the mechanics could probably take a long time. My other fear is that if votes stretch out it will just go forgotten and we'll circle around endlessly. Since voting is (as you put in a comment above) a fall-back, I think it's reasonable to say that voting should be in reaction to a thorough discussion, rather than before one.

Author: I'm fine making a vote take 15 days & keeping this at 5. A standard review is 3 people. A vote-worthy disagreement means that you already have 3 votes. 5 votes means at least two more people need to look into an issue. That said, given my comment above, it would be nice to be able to cut a 15 day voting window short. I have thoughts on criteria to cut votes short, but they might not be worth the effort, since most things in tedana aren't THAT rushed.

Reviewer: I think it's not worth the effort.

   f) The outcome of the vote is decided based on a simple majority.

Steering Committee
```````````````````
The steering committee steers. The goal of the steering committee is to help
guide the direction of the project. Decisions in the steering committee will 
focus on how to present project issues to the broader community in a clear way
rather than making project decisions without community input. 

The steering committee can decide:

> - An issue should be prioritized for wider communal discussion 

Reviewer: ```suggestion
- An issue should be prioritized for wider communal discussion.
```

- A a pull request requires more discussion or reviews than standard before
  merging.
- How a breaking change (something that changes existing user function calls
  or program outputs) will be presented to the developer and user base for
  discussion, before decisions are made.

> - Criteria for cutting a new version release and when those criteria are met

Reviewer: ```suggestion
- Criteria for cutting a new version release and when those criteria are met.
```

  community. Maintainers have additional authority (see `Decision Making Process`_) helping them to
  resolve conflicts and increase the pace of the development when necessary.
  Any maintainer can remove themselves. Any contributor can become a
  maintainer by request, or by nomination of a current maintainer,  and with the support of the majority of the current
  maintainers. Current Maintainers:

  +-----------------------------------+
  | `Logan Dowdle`_ (@dowdlelt)       | 
  +-----------------------------------+
  | `Elizabeth DuPre`_ (@emdupre)     |
  +-----------------------------------+
  | `Dan Handwerker`_ (@handwerkerd)  |
  +-----------------------------------+
  | `Taylor Salo`_ (@tsalo)           |
  +-----------------------------------+
  | `Joshua Teves`_ (@jbteves)        |
  +-----------------------------------+
  | `Eneko Uruñuela`_ (@eurunuela)    |
  +-----------------------------------+
  | `Kirstie Whitaker`_ (@KirstieJane)|
  +-----------------------------------+

Steering committee
  The steering committee is made up of a subset of maintainers who help guide
  the project. Current Steering Committee members:

>   +----------------------------------------------+
>   | `Dan Handwerker`_ (@handwerkerd) (tentative) |
>   +----------------------------------------------+ 
>   | Your name here!                              |
>   +----------------------------------------------+
>   | Your name here!                              |
>   +----------------------------------------------+
>   | Your name here!                              |
>   +----------------------------------------------+
>   | Your name here!                              |
>   +----------------------------------------------+

Author: ```suggestion
  +--------------------------------------+
  | `Logan Dowdle`_ (@dowdlelt)              | 
  +--------------------------------------+
  | `Elizabeth DuPre`_ (@emdupre)           |
  +--------------------------------------+
  | `Dan Handwerker`_ (@handwerkerd)   |
  +--------------------------------------+
  | `Taylor Salo`_ (@tsalo)                           |
  +--------------------------------------+
  | `Joshua Teves`_ (@jbteves)                  |
  +--------------------------------------+
  | `Eneko Uruñuela`_ (@eurunuela)          |
  +--------------------------------------+
```

project to overly rely on the existence of one person. We developed the
following system with several goals in mind:

- Grow the community.
- Strive for consensus.
- Provide a path for when consensus cannot be achieved.
- Minimize the administrative burden.
- Maximize the `bus factor`_ of the project.
- Acknowledge the leadership and time multiple people contribute to our
  community without demanding more time than any individual can offer.
  Dividing leadership responsabilies into multiple smaller roles also
  makes it easier to encourage new people to take on a leadership role
  without fearing that too much work will be required of them.
- Openness as a priority:

  - Promote open discussions.
  - Openness is critical to building trust with the broader community
  - Openness provides a mechanism for non-leaders to identify and address
    oversights or mistakes
  - Openness provides a smoother transition to onboard future leaders
  - Leadership meetings should be open and notes should be shared unless
    there are discussions about sensitive personal matters.

This governance structure is a work-in-progress. We welcome both people

> who want to take on a leadership role as well as ideas for our to improve

Reviewer 5: ```suggestion
who want to take on a leadership role as well as ideas to improve
```

  A contributor is someone who has made a contribution to tedana. A
  contribution, can be code, documentation, or conceptual. All contributors are
  listed in the `all-contributors file`_. The community decides on the content
  of this file using the same process as any other change to the `Repository`_
  (see below) allowing the meaning of "Contributor" to evolve independently of
  the Decision-making rules. Contributors also have the option to be added to
  the Zenodo file which may be used for authorship credit for tedana.
    

Maintainer
  A Contributor is responsible for the long term health of the project and the
  community. Maintainers have additional authority (see `Decision Making Process`_) helping them to
  resolve conflicts and increase the pace of the development when necessary.
  Any maintainer can remove themselves. Any contributor can become a
  maintainer by request, or by nomination of a current maintainer,  and with the support of the majority of the current
  maintainers. Current Maintainers:

  +-----------------------------------+
  | `Logan Dowdle`_ (@dowdlelt)       | 
  +-----------------------------------+
  | `Elizabeth DuPre`_ (@emdupre)     |

>   +-----------------------------------+

Author: ```suggestion
  +-----------------------------------+
  | `Javier Gonzalez-Castillo`_ (@javiergcas) |
  +-----------------------------------+
```

  | `Dan Handwerker`_ (@handwerkerd)  |
  +-----------------------------------+
  | `Taylor Salo`_ (@tsalo)           |
  +-----------------------------------+
  | `Joshua Teves`_ (@jbteves)        |
  +-----------------------------------+
  | `Eneko Uruñuela`_ (@eurunuela)    |
  +-----------------------------------+
  | `Kirstie Whitaker`_ (@KirstieJane)|
  +-----------------------------------+

> Steering committee

Reviewer 5: Maybe you could add a link to the _Steering Committee_ section below.

Reviewer: Added in ff7f48b

  - | Processing algorithms leader `Eneko Uruñuela`_ (Decomposition) & `Dan Handwerker`_ (Metrics & Decision Tree)
    |   Someone who can make sure algorithms are appropriately implemented (or knows enough to delegate to someone who can make sure implementation is good)
    |   Someone who can either answer processing algorithm questions or direct people to where they can find answers
  - | Collaborative programming leader `Elizabeth DuPre`_ & `Joshua Teves`_
    |   Helps make sure tedana is following best practices for Python code design, testing, and communication for issues/pull requests etc.
  - | Communications leader `Joshua Teves`_
    |   Mailing list manager & other outward-facing communication about the project
  - | New contributors leader `Taylor Salo`_
    |   Leads efforts to make contributor documentation more welcoming
    |   Is a point of contact for potential contributors to make them feel welcome and direct them to relevant resources or issues
  - | Multi-echo fMRI support leader `Logan Dowdle`_
    |   Monitors places where people may ask questions about tedana or multi-echo fMRI and tries to find someone to answer those questions

>   - | Enforcer(s) of the `code of conduct`_ `Elizabeth DuPre`_ & `Dan Handwerker`_ & another volunteer

Reviewer 5: Are we adding @smoia here?

Reviewer: @smoia if you're willing to that would be great, could you confirm that you'd be interested?

```````````````````
The steering committee steers. The goal of the steering committee is to help
guide the direction of the project. Decisions in the steering committee will 
focus on how to present project issues to the broader community in a clear way
rather than making project decisions without community input. 

The steering committee can decide:

- An issue should be prioritized for wider communal discussion.
- A a pull request requires more discussion or reviews than standard before
  merging.
- How a breaking change (something that changes existing user function calls
  or program outputs) will be presented to the developer and user base for
  discussion, before decisions are made.
- Criteria for cutting a new version release and when those criteria are met.

Steering committee decisions should strive for consensus. If consensus cannot
be reached, the members of the steering committee should vote. Voting will take
place over 7 days or until every steering committee member votes or abstains.
The outcome of a vote is based on a simple majority.

.. _César Caballero-Gaudes: https://github.com/CesarCaballeroGaudes
.. _Logan Dowdle: https://github.com/dowdlelt

> .. _Elizabeth DuPre: https://github.com/emdupre

Author: ```suggestion
.. _Elizabeth DuPre: https://github.com/emdupre
.. _Javier Gonzalez-Castillo: https://github.com/javiergcas
```

Governance
==========

> 

Reviewer 2: ```suggestion
Governance is a hugely important part of any project.
It is especially important to have clear process and communication channels
for open source projects that rely on a distributed network of volunteers, such as ``tedana``.
```

- Strive for consensus.
- Provide a path for when consensus cannot be achieved.
- Grow the community.
- Maximize the `bus factor`_ of the project.
- Acknowledge the leadership and time multiple people contribute to our
  community without demanding more time than any individual can offer.
  Dividing leadership responsabilies into multiple smaller roles also
  makes it easier to encourage new people to take on a leadership role
  without fearing that too much work will be required of them.

This governance structure is a work-in-progress. We welcome both people
who want to take on a leadership role as well as ideas for our to improve
this structure.

Steering Committee and Leadership Roles
---------------------------------------

Organizational Principles
`````````````````````````
The steering committee steers. The goal of the steering committee is to help
guide the direction of the project. Decisions in the steering committee will 
focus on how to present project issues to the broader community in a clear way
rather than making project decisions without community input. 

Issues or pull requests that can be reviewed by any community member can be

> reviewed and approved by the steering committee or steering committee members.

Reviewer 2: I'm not sure that I understand this point ? Are we just trying to say that steering committee members do not have special approval above other community members ?

Technically, @tsalo and I have admin powers and we have _occasionally_ used them, although we try to avoid doing so as much as possible. I'd imagine steering committee members will likely have similar powers / process, no ?

Author: As it's written now, the steering committee has the power to make a PR more difficult to approve and to decide the best way to engage the broader community for significant PRs. Maintainers have the ability to decide to approve something that doesn't meet the standard criteria. In practice that means someone with admin powers should consult the maintainers before using those powers.
I just copied from the governance doc that we had been editing. This issue may be because of a disconnect caused by me writing the steering section and Taylor focusing on Maintainers in the governance section. Perhaps some of the powers Taylor gave to maintainers should rest with the steering committee. I'll read through this all again and welcome thoughts from others as well.

The steering committee can decide that an issue should be prioritized for wider
communal discussion or that a pull request requires more discussion or reviews
than standard before merging. The steering committee can also decide how a
breaking change (something that changes existing user function calls or program
outputs) will be presented to the developer and user base for discussion, before
decisions are made.

Steering committee decisions should strive for consensus and accept the

> majority’s decision. If voting is necessary, it can happen asynchronously and is

Reviewer 2: Which majority ? The following sentences makes it seem like this is an internal steering committee vote, rather than a community vote, but it's not completely clear to me.

Author: The idea is that the steering committee has the power to decide how we engage with the broader community and get feedback in situations where communal voices are necessary for the decision process. I agree we can be more specific, but I'll need to think about how to add specificity without **unnecessarily** reducing administrative flexibility

defined as more than half of the members of the steering committee. If a member
of the steering committee is unable to spend the time necessary to understand
an issue & vote, they can recuse from a decision so that the number of committee

> members is reduced

Reviewer 2: ```suggestion
members is reduced.
```

Openness is a priority:

- Openness is critical to building trust with the broader community
- Openness provides a mechanism for non-steering committee members to identify
  and address steering committee blindspots
- Openness provides a smoother transition to onboard future steering committee
  members
- If steering committee discussions (written or verbal) could include

>   community members without compromising efficiency, should be invited. The

Reviewer: ```suggestion
  community members without compromising efficiency, they should be invited. The
```

  steering committee can schedule discussions without needing to ask about the
  availability for people outside of the steering committee. If notes from
  meetings can be openly shared without compromising personal privacy, they

>   should be.

Reviewer 2: I might invert this to highlight that whenever possible, steering committee meetings will be open and minutes will be freely available. There are, of course, some circumstances when this will not be possible, but I think it's a reasonable default !

The current members of the tedana steering committee are:

- Name 1 & link to github profile
- Name 2 & link to github profile
- Name 3 & link to github profile
- Name 4 & link to github profile

> - Name 5 & link to github profile

Reviewer 2: Is this the maintainer list, or do you see this as a different group ?

Author: The steering committee is separate from the maintainers. The maintainer list has no size limit. The steering committee is a smaller group that can more efficiently communicate & meet. I (we?) envisioned these as two separate groups, but several questions above make me wonder if they should be. We've effectively separated the leadership roles from steering committee membership. Does that mean our group of maintainers is effectively the steering committee? If there are ~10 maintainers is that too many people to commit to be involved in all decisions?

Also, one of your comments above mentioned Admin privileges. We don't talk at all about any leader having those privileges or the appropriate situations to use them. That seems like an oversight.

Reviewer 2: Before reading this, I was honestly under the impression that maintainers and steering were the same. I thought we had discussed having representation across the different areas on steering--hence the time spent defining them--but I'm not sure if I had just conflated the two ideas.

Either way, I think a smaller steering makes sense (I'd suggest 6 members). It would also make sense to me to extend that to those folks who regularly come to developer calls, since they're already very familiar with the project and have a history of involvement.

Author: Current proposal:
- Leaders can be anyone
- Maintainers are people who have a standing commitment to contribute
- The steering committee is a subset of maintainers
- We still need to clarify the specific things that the steering committee can do that maintainers cannot
It would be good to get @tsalo's thoughts on this.

If you are interested in joining the steering committee, please let the current members know.

Leadership roles
````````````````
We have identified key skills that help advance tedana development and
volunteers who will be the point-people for each of these roles. One person
can fill more than one role and more than one person can decide to share or
split the responsabilities of a role. The steering committee will include
people who also take on leadership roles, but a person can take on a leadership

> role without also volunteering to be on the steering committee.

Reviewer 2: Ah I see. Should this section go before `steering committee`, then, since folks will more directly work with those in leadership roles, and they're also easier to transition into ?

  |   If this is one person, that person should NOT be on the steering committee
  |   If this is more than one person, at least one should not be on the steering committee
  |   Ideal is someone who cares about tedana but DOESN’T know contributors well enough to say, ”Person X would never do that”

Changing leaders
````````````````
Steering committee members can remove themselves from the steering committee at
any time and open up a call for a new self-nomination. Anyone can request to take
on a leadership role at any time. Once per year, there should be an explicit call
to the larger contributor community asking if anyone wants to self nominate for
membership on the steering committee or other leadership roles. If individuals
cannot reach consensus on who steps back and who assumes new roles, then a
majority vote of contributors from the previous 3 years will assign people to
roles where there are conflicts.

If there are concerns with a tedana steering committee member or leader, any
enforcer of the code of conduct can ask anyone to step down from a leadership role.
If a person refuses to step down, then an enforcer of the code of conduct can call

> a vote of contributors to remove an individual from a leadership role in tedana.

Reviewer 2: Should it be a vote of all contributors ? I worry that this could mean putting out really personal information, depending on the code of conduct violation. I'd rather consensus among the CoC enforcers.

Author: I commented on this in the original doc as well, but I don't know the sweet spot. In an ideal world, a code of conduct enforcer will confront someone & ask them to step aside & they'd rather step aside than cause additional strife in the project. Forcing someone out only becomes an issue if the person refuses. If the forcing out process is limited to any subset of leaders then there's the risk of people sticking with their friend rather than the accuser. If we open it up completely, then it can be a privacy violation. Maybe we can give the accuser the power to escalate from CoC enforcers to Steering Committee to Contributors if they aren't believed in the smaller groups?

Reviewer 2: I can't think of another organization where CoC disputes are brought to a public vote. I think it could turn out really badly for everyone, especially the victim. Honestly, I worry that folks would be less likely to report if that was even a possible outcome.

I understand your concern about viable checks, but I'd suggest this decision / discussion is left to the CoC enforcement team (and future PRs). Here, I'd just point to the enforcement team.

Reviewer 2: I might even leave out the bit about a refusal to step down. That can be detailed in an Enforcement section of the CoC as it's agreed upon. I do think we could clearly list here, though, who the current Enforcement team includes.

Author: 💯 agree with your concerns. I do want a clear Code of Conduct enforcement path if a very core project leader or a Code of Conduct enforcer is the violator. Those are the situations that tend to become the most ugly. I **really** hope this doesn't become an issue given the proposed CoC enforcers are me & @emdupre, but this is something that's worth future-proofing. I wonder there are existing models for CoC enforcement that account for these scenarios in for small projects like tedana. @KirstieJane?

`lazy consensus system used in the Apache Foundation <https://www.apache.org/foundation/voting.html>`_,
and heavily depend on the
`GitHub Pull Request review system <https://help.github.com/articles/about-pull-requests/>`_.

Definitions
```````````

Repository
  `https://github.com/ME-ICA/tedana <https://github.com/ME-ICA/tedana>`_

Contributor
  Person listed in the `all-contributors file`_.
  The community decides on the content of this file using the same process as any
  other change to the Repository (see below) allowing the meaning of "Contributor"
  to evolve independently of the Decision-making rules.

Maintainer
  A Contributor responsible for the long term health of the project and the
  community. Maintainers have additional rights (see Rules) helping them to
  resolve conflicts and increase the pace of the development when necessary.
  Any maintainer can self-remove themselves. Any contributor can become a
  maintainer by request and with the support of the majority of the current
  maintainers. Current Maintainers:

>   +-----------------------------------+-----------------+

Reviewer: I'm not convinced we should retain this table. It could be aggravating to track.

Reviewer 2: I also worry that time commitments are likely to fluctuate quite a bit over the year, etc... since no one is paid, it's hard to guarantee these times. Both in a good (some weeks I feel like I'm closer to a full hour, or more !) and a bad (other weeks I'm definitely at zero hours) way.
So, I'm +1 on dropping the table.

Author: I'm neutral on the table. I just copied it from the design doc. The benefit is it does let new people know that we're all pushing this project forward by doing our small parts. The negative is that the numbers will almost always be wrong. One compromise is set it to a monthly or quarterly commitment rather than a weekly commitment. That is Dan spends 5-10h/quarter

Reviewer 2: That might make the numbers (more or less) correct, but thinking on it more, I'm still worried the table sets the wrong impression (ie, that these folks will work on this for X hours). Most of our development is still done by a volunteer team, and I think it's important prominently acknowledge that.

I'm wondering if this table was inspired from BIDS. They pay their maintainers, so it would make more sense for them.

  | Name                              | Time commitment |
  +===================================+=================+
  | `Logan Dowdle`_ (@dowdlelt)       | 0.5h/week       |
  +-----------------------------------+-----------------+
  | `Elizabeth DuPre`_ (@emdupre)     | 0.5h/week       |
  +-----------------------------------+-----------------+
  | `Dan Handwerker`_ (@handwerkerd)  | 0.5h/week       |
  +-----------------------------------+-----------------+
  | `Ross Markello`_ (@rmarkello)     | 0.5h/week       |
  +-----------------------------------+-----------------+
  | `Taylor Salo`_ (@tsalo)           | 3h/week         |
  +-----------------------------------+-----------------+
  | `Joshua Teves`_ (@jbteves)        | 0.5h/week       |
  +-----------------------------------+-----------------+
  | `Eneko Uruñuela`_ (@eurunuela)    | 0.5h/week       |
  +-----------------------------------+-----------------+
  | `Kirstie Whitaker`_ (@KirstieJane)| 0.5h/week       |
  +-----------------------------------+-----------------+

Rules
`````

1. Potential modifications to the Repository should first be proposed via an 

>    Issue. Rules regarding Votes apply to both Pull Requests and Issues.

Reviewer 2: Bit confusing, since we've barely introduced voting yet. Could we cross-ref it, at least ? Or did I miss an earlier introduction of the role of voting ?

   - Every modification of the specification (including a correction of a typo, adding a new Contributor, an extension adding support for a new data type, or others) or proposal to release a new version needs to be done via a Pull Request (PR) to the Repository.
2. Anyone can open a PR (this action is not limited to Contributors).
3. PRs adding new Contributors must also add their GitHub names to the 

>    `all-contributors file`_. 

Reviewer 2: I don't think we want anyone to edit that file directly !!!! Only the bot should do it.
I'd rather that we _offer_ (not force) folks to add their info to the Zenodo, and in that PR we can tag the bot to add them to the table.

Author: I think I might just remove this bullet point about PRs for new contributors. We're already saying elsewhere that contributors should be in the all-contributors file and that contrbutors are one thing that's added through PR. Why exactly do we need a specific rule that is seems overly detailed for this doc?

   This should be done with the allcontributors bot.

   - Contributors may also add themselves to the Zenodo file if they wish, but this is not mandatory.
4. A PR is eligible to be merged if and only if these conditions are met:

   a) The PR features at least two `Reviews that Approve <https://help.github.com/articles/about-pull-request-reviews/#about-pull-request-reviews>`_

>       the PR from Maintainers of which neither is the author of the PR. 

Reviewer 2: Do the approvals _have_ to be from Maintainers ? Personally, I'm fine if approvals come from any community member.

Reviewer: I'm not wild about any community member approving stuff, I would feel more warm and cozy knowing that maintainers are approving. Honestly, though, I just doubt that difference will come up much. Anyone who's approving reviews often and doing a decent job will de facto become a maintainer. So perhaps we should go with the more open model of community members being allowed to approve.

Author: I'd rather give anyone the option to review & approve. We've already had one case on BIDS derivatives where we asked for a BIDs developer who isn't non-tedana person to review. (He didn't, but it's still a use case for this scenario).
With the proposed system, the steering committee can raise the bar for any PR so I'd rather keep it so that the default is any contributor can review, but the steering committee can change that.

Author: Also, given how long it takes for our current Maintainers to review some of the harder PRs, if an enthusiastic contributor comes along and wants to do a serious review, I'm all for it!

>       The reviews need to be made after the last commit in the PR (equivalent to 

Reviewer: I believe we're considering amending this. Thoughts, @tsalo and @emdupre ?

Reviewer 2: Yeah, it gets a bit frustrating when we have "and one more thing..." which I am incredibly guilty of.
The spirit of having two approving reviews is what's important, really -- we don't want someone to have reviewed three months ago, the PR had a lot of changes, and then only one more approval comes in and it's merged. I'm flexible on how to communicate / enforce that, though.

Reviewer 4: Agreed. We should find a way to enforce the spirit of this without having to re-review after fixing typos. In BIDS, they do a pretty good job of keeping approvals after changes without letting major changes slip through, so there must be a way. I say we change it to 

> The reviews need to be made after all major changes have been made in the PR. If new major changes are introduced, reviews must be resubmitted. This does not apply to minor changes, like typo corrections.

And we remove the "discard stale reviews" rule in the repo settings.

      `Stale review dismissal <https://help.github.com/articles/enabling-required-reviews-for-pull-requests/>`_
      option on GitHub).
   b) Does not feature any `Reviews that Request changes <https://help.github.com/articles/about-required-reviews-for-pull-requests/>`_.

>    c) Does not feature "WIP" in the title (Work in Progress).

Reviewer: We have moved more towards Draft PRs, like this one. I think we should amend this language to reflect that.

Reviewer 2: +1 on draft PRs.

   d) Passes all automated tests.

>    e) Is not proposing a new release or has been approved by at least one

Reviewer 2: I would think only steering committee members can cut releases ? WDYT ?

Reviewer: I agree @emdupre

Reviewer: Resolved as of latest commit.

      Maintainer (i.e., PRs proposing new releases need to be approved by
      at least one Maintainer).
5. After consultation with contributors, the steering committee can decide
   to merge any PR - even if it's not eligible to merge according to Rule 4.

> 6. Any Maintainer can Review a PR and request changes. If a Maintainer Requests
>    changes they need to provide an explanation regarding what changes should
>    be added and justification of their importance. Reviews requesting
>    changes can also be used to request more time to review a PR.

Reviewer 2: ```suggestion
6. Any community member can Review a PR and request changes. If a community member Requests
   changes they need to provide an explanation regarding what changes should
   be added and justification of their importance. Reviews requesting
   changes can also be used to request more time to review a PR.
```

> 7. A Maintainer who Requested changes can Dismiss their own review or Approve

Reviewer 2: ```suggestion
7. A community member who Requested changes can Dismiss their own review or Approve
```

   changes added by the Contributor who opened the PR.

> 8. If the author of a PR and Maintainer who provided Review that Requests

Reviewer: I'm not entirely sure I follow this rule; could you edit for clarity @handwerkerd ?

Reviewer 2: I agree, I'm also a bit lost here.

Author: @tsalo write this and can hopefully explain more. I assume it means that a maintainer is allowed to change their mind if they asked for a change that they later decided wasn't necessary.

Reviewer 4: This is one of the rules I took from BIDS. My understanding is that, if a maintainer is being intractable and requesting unreasonable changes, the rest of the maintainers can override their review based on a vote.

Governance
==========
Governance is a hugely important part of any project.

> It is especially important to have clear process and communication channels

Reviewer 2: ```suggestion
It is especially important to have clear processes and communication channels
```

for open source projects that rely on a distributed network of volunteers,
such as ``tedana``.

Overview
--------

Tedana is a relatively small open source project that requires specialized
knowledge in multiple domains.
This leads to several challenges.
No one
person on the current tedana development team has a combination of
available time plus expertise in collaborative software development, MRI
physics, and advanced data processing methods to assume a primary project
leader role.
Even if such a person was interested, it may not benefit the
project to overly rely on the existence of one person.

> We developed the

Reviewer 2: ```suggestion
Instead, we developed the
```

following system with several goals in mind:

- Grow the community.
- Strive for consensus.
- Provide a path for when consensus cannot be achieved.
- Minimize the administrative burden.
- Maximize the `bus factor`_ of the project.
- Acknowledge the leadership and time multiple people contribute to our
  community without demanding more time than any individual can offer.

>   Dividing leadership responsabilities into multiple smaller roles also

Reviewer: ```suggestion
  Dividing leadership responsibilities into multiple smaller roles also
```

- Openness as a priority:

  - Promote open discussions.
  - Openness is critical to building trust with the broader community
  - Openness provides a mechanism for non-leaders to identify and address
    oversights or mistakes
  - Openness provides a smoother transition to onboard future leaders
  - Leadership meetings should be open and notes should be shared unless
    there are discussions about sensitive personal matters.

This governance structure is a work-in-progress.
We welcome both people
who want to take on a leadership role as well as ideas to improve
this structure.

Leadership
----------

Contributor
  A contributor is someone who has made a contribution to tedana.
  A contribution can be code, documentation, or conceptual.
  All contributors are listed in the `all-contributors file`_.
  The community decides on the content of this file using the same process
  as any other change to the `Repository`_ (see below) allowing the
  meaning of "Contributor" to evolve independently of the Decision-making
  rules.
  Contributors also have the option to be added to the Zenodo file which
  may be used for authorship credit for tedana.

Maintainer

>   A Maintiner is responsible for the long term health of the project and

Reviewer: ```suggestion
  A Maintainer is responsible for the long term health of the project and
```

   d) Passes all automated tests.
   e) Is not proposing a new release.
   f) The steering committee has not added extra restrictions.
      For example, if a PR is a non-trival change, the steering committee
      can create a system to get feedback from more than just two reviewers
      before merging.
5. After consultation with contributors, the steering committee can decide
   to merge any PR - even if it's not eligible to merge according to Rule 4.
6. Anyone can Review a PR and request changes.
   If a community member requests changes they need to provide an
   explanation regarding what changes should be made and justification of
   their importance.
   Reviews requesting changes can also be used to request more time to
   review a PR.
7. A reviewer who requested changes can dismiss their own review, if they
   decide their requested changes are no longer necessary, or approve
   changes that address the issue underlying their change request.
8. If the author of a PR and a reviewer who requests changes cannot find a
   solution that would lead to:

   (1) The author closing the PR without merging
   (2) The reviewer accepting requested changes or

>    (3) The dismissing their review, so that the PR can be approved and

Reviewer 4: missing info

Author: What's missing?

Reviewer 4: The person doing the dismissing isn't clear.

Reviewer 2: I suggested it should be the reviewer dismissing their own review. Let us know if that's not what you were thinking @handwerkerd  ?

>    (3) The dismissing their review, so that the PR can be approved and

Reviewer 2: ```suggestion
   (3) The reviewer dismissing their review, so that the PR can be approved and
```

>    (3) The dismissing their review, so that the PR can be approved and

Author: ```suggestion
   (3) The reviewer dismissing their own review, so that the PR can be approved and
```

      vote, the Steering Committee has the authority to create a voting
      process to help resolve disagreements in a more efficient and
      respectful manner.
   b) Only Contributors can vote and each Contributor gets one vote.
   c) A vote ends after 15 working days or when all Contributors have
      voted or abstained (whichever comes first).
   d) A vote freezes the PR - no new commits or Reviews Requesting Changes
      can be added to it while a vote is ongoing.
      If a commit is accidentally made during that period it should be
      reverted.
      Comments are allowed.
   e) The quorum for a vote is five votes.
   f) The outcome of the vote is decided based on a simple majority.

.. _Steering Committee:

Steering Committee
```````````````````
The steering committee steers.
The goal of the steering committee is to help guide the direction of the
project.
Decisions in the steering committee will focus on how to present project
issues to the broader community in a clear way rather than making project
decisions without community input.

The steering committee can decide:

- An issue should be prioritized for wider communal discussion.

> - A a pull request requires more discussion or reviews than standard before

Reviewer 4: ```suggestion
- A pull request requires more discussion or reviews than standard before
```

physics, and advanced data processing methods to assume a primary project
leader role. Even if such a person was interested, it may not benefit the
project to overly rely on the existence of one person. We developed the

> following system the several goals in mind:

Reviewer 4: ```suggestion
following system with several goals in mind:
```
